pkg/reconciler/pipelinerun/resources/apply.go

```diff
@@ -150,13 +150,63 @@ func ApplyContexts(spec *v1.PipelineSpec, pipelineName string, pr *v1.PipelineRu
 	return ApplyReplacements(spec, GetContextReplacements(pipelineName, pr), map[string][]string{}, map[string]map[string]string{})
 }
 
+// filterMatrixContextVar returns a list of params which contain any matrix context variables such as
+// $(tasks.<pipelineTaskName>.matrix.length) and $(tasks.<pipelineTaskName>.matrix.<resultName>.length)
+func filterMatrixContextVar(params v1.Params) v1.Params {
+	var filteredParams v1.Params
+	for _, param := range params {
+		if expressions, ok := param.GetVarSubstitutionExpressions(); ok {
+			for _, expression := range expressions {
+				// tasks.<pipelineTaskName>.matrix.length
+				// tasks.<pipelineTaskName>.matrix.<resultName>.length
+				subExpressions := strings.Split(expression, ".")
+				if subExpressions[2] == "matrix" && subExpressions[len(subExpressions)-1] == "length" {
+					filteredParams = append(filteredParams, param)
+				}
+			}
+		}
+	}
+	return filteredParams
+}
+
 // ApplyPipelineTaskContexts applies the substitution from $(context.pipelineTask.*) with the specified values.
-// Uses "0" as a default if a value is not available.
-func ApplyPipelineTaskContexts(pt *v1.PipelineTask) *v1.PipelineTask {
+// Uses "0" as a default if a value is not available as well as matrix context variables
+// $(tasks.<pipelineTaskName>.matrix.length) and $(tasks.<pipelineTaskName>.matrix.<resultName>.length)
+func ApplyPipelineTaskContexts(pt *v1.PipelineTask, pipelineRunStatus v1.PipelineRunStatus) *v1.PipelineTask {
 	pt = pt.DeepCopy()
+	var pipelineTaskName string
+	var resultName string
+	var matrixLength int
+
 	replacements := map[string]string{
 		"context.pipelineTask.retries": strconv.Itoa(pt.Retries),
 	}
+
+	filteredParams := filterMatrixContextVar(pt.Params)
+
+	for _, p := range filteredParams {
+		pipelineTaskName, resultName = p.ParseTaskandResultName()
+		// find the referenced pipelineTask to count the matrix combinations
+		if pipelineTaskName != "" && pipelineRunStatus.PipelineSpec != nil {
+			for _, task := range pipelineRunStatus.PipelineSpec.Tasks {
+				if task.Name == pipelineTaskName {
+					matrixLength = task.Matrix.CountCombinations()
+				}
+			}
+			replacements["tasks."+pipelineTaskName+".matrix.length"] = strconv.Itoa(matrixLength)
+		}
+		// the number of child refs will map to the number of results produced by the matrixed PipelineTask
+		if pipelineTaskName != "" && resultName != "" {
+			var resultsLength int
+			for _, ref := range pipelineRunStatus.ChildReferences {
+				if ref.PipelineTaskName == pipelineTaskName {
+					resultsLength++
```

Comment:
We need to consider a scenario here that a result is not initialized by one of the instances. 

After retrieving the name of the taskRun through childReferences, we need to check if a task result was initialized in that taskRun before incrementing the resultsLength counter.

I am looking if there is any other alternate better way of retrieving this.

Reply:
We can maintain a cache of results like we had discussed before, else this might require a `Get` as a last resort:

```
c.PipelineClientSet.TektonV1().TaskRuns(pr.Namespace).Get(ctx, tr, metav1.GetOptions{})
```

Reply:
Can we utilize - https://github.com/tektoncd/pipeline/pull/7167/files#diff-39d9baa708e379638f742c8b9fd36d9d15c64c3386de958b5a3f309ca8fa037aR749?